src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,456 @@
+
+            all_params.append(dict(self._policy.named_parameters()))
+            # Restore to pre-updated policy
+            update_module_params(self._policy, theta)
+
+        return all_samples, all_params
+
+    def _adapt(self, itr, batch_samples, set_grad=True):
+        """Performs one MAML inner step to update the policy.
+
+        Args:
+            itr (int): Iteration.
+            batch_samples (MAMLTrajectoryBatch): Samples data for one
+                task and one gradient step.
+            set_grad (bool): if False, update policy parameters in-place.
+                Else, allow taking gradient of functions of updated parameters
+                with respect to pre-updated parameters.
+
+        """
+        # pylint: disable=protected-access
+        loss = self._inner_algo._compute_loss(itr, *batch_samples)
+
+        # Update policy parameters with one SGD step
+        self._inner_optimizer.zero_grad()
+        loss.backward(create_graph=set_grad)
+
+        with torch.set_grad_enabled(set_grad):
+            self._inner_optimizer.step()
+
+    def _meta_optimize(self, itr, all_samples, all_params):
+        if isinstance(self._meta_optimizer, ConjugateGradientOptimizer):
+            self._meta_optimizer.step(
+                f_loss=lambda: self._compute_meta_loss(
+                    itr, all_samples, all_params, set_grad=False),
+                f_constraint=lambda: self._compute_kl_constraint(
+                    itr, all_samples, all_params))
+        else:
+            self._meta_optimizer.step(lambda: self._compute_meta_loss(
+                itr, all_samples, all_params, set_grad=False))
+
+    def _compute_meta_loss(self, itr, all_samples, all_params, set_grad=True):
+        """Compute loss to meta-optimize.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): A two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            all_params (list[dict]): A list of named parameter dictionaries.
+                Each dictionary contains key value pair of names (str) and
+                parameters (torch.Tensor).
+            set_grad (bool): Whether to enable gradient calculation or not.
+
+        Returns:
+            torch.Tensor: Calculated mean value of loss.
+
+        """
+        theta = dict(self._policy.named_parameters())
+        old_theta = dict(self._old_policy.named_parameters())
+
+        losses = []
+        for task_samples, task_params in zip(all_samples, all_params):
+            for i in range(self._num_grad_updates):
+                self._adapt(itr, task_samples[i], set_grad=set_grad)
+
+            update_module_params(self._old_policy, task_params)
+            with torch.set_grad_enabled(set_grad):
+                # pylint: disable=protected-access
+                loss = self._inner_algo._compute_loss(itr, *task_samples[-1])
```

Comment:
It would be nice to change the inner algorithms to make it clear this is part of the public API.

Reply:
@ryanjulian What do you think?

Reply:
yes -- we can focus on this later